source/Locales/BooleanAlgebra.lagda

```diff
@@ -0,0 +1,719 @@
+                               → (x y : ⟪ B ⟫)
+                               → (x ≤[ poset-of-ba B ] y
+                               ↔ η x ≤[ poset-of L ] η y) holds
+embedding-is-order-isomorphism B L η μ x y = † , ‡
+ where
+  η-meet-preserving : (x y : ⟪ B ⟫) → η (x ⋏[ B ] y) ＝ η x ∧[ L ] η y
+  η-meet-preserving = embedding-preserves-meets B L η μ
+
+  † : (x ≤[ poset-of-ba B ] y ⇒ η x ≤[ poset-of L ] η y) holds
+  † p = η x              ＝⟨ ap η (※ ⁻¹)              ⟩ₚ
+        η (x ⋏[ B ] y)   ＝⟨ η-meet-preserving x y    ⟩ₚ
+        η x ∧[ L ] η y   ≤⟨ ∧[ L ]-lower₂ (η x) (η y) ⟩
+        η y              ■
+   where
+    open PosetReasoning (poset-of L)
+
+    ※ : x ⋏[ B ] y ＝ x
+    ※ = ≤-is-antisymmetric (poset-of-ba B) ※₁ ※₂
+     where
+      ※₁ : ((x ⋏[ B ] y) ≤[ poset-of-ba B ] x) holds
+      ※₁ = ⋏[ B ]-is-lower₁ x y
+
+      ※₂ : (x ≤[ poset-of-ba B ] (x ⋏[ B ] y)) holds
+      ※₂ = ⋏[ B ]-is-greatest (≤-is-reflexive (poset-of-ba B) x) p
+
+  ‡ : (η x ≤[ poset-of L ] η y ⇒ x ≤[ poset-of-ba B ] y) holds
+  ‡ p = x ＝⟨ ♠ ⁻¹ ⟩ₚ x ⋏[ B ] y ≤⟨ ⋏[ B ]-is-lower₂ x y ⟩ y ■
+   where
+    open PosetReasoning (poset-of-ba B)
+
+    ♥ : η (x ⋏[ B ] y) ＝ η x
+    ♥ = η (x ⋏[ B ] y)     ＝⟨ embedding-preserves-meets B L η μ x y ⟩
+        η x ∧[ L ] η y     ＝⟨ connecting-lemma₁ L p ⁻¹              ⟩
+        η x                ∎
+
+    ♠ : x ⋏[ B ] y ＝ x
+    ♠ = embedding-injective B L η μ (x ⋏[ B ] y) x ♥
+
+embeddings-lemma : (B : BooleanAlgebra 𝓤′ 𝓥′) (L : Frame 𝓤 𝓥 𝓦)
+                 → (η : ⟪ B ⟫ → ⟨ L ⟩)
+                 → is-embedding B L η holds
+                 → (x : ⟪ B ⟫) → (η x ≤[ poset-of L ] 𝟎[ L ]) holds → x ＝ ⊥[ B ]
+embeddings-lemma B L η (ι , _ , (_ , ξ , _)) x p = ι x ⊥[ B ] †
+ where
+  † : η x ＝ η ⊥[ B ]
+  † = η x ＝⟨ only-𝟎-is-below-𝟎 L (η x) p ⟩ 𝟎[ L ] ＝⟨ ξ ⁻¹   ⟩ η ⊥[ B ] ∎
+
+\end{code}
+
+\begin{code}
+
+is-generated-by : (L : Frame 𝓤 𝓦 𝓦) → (B : BooleanAlgebra 𝓦 𝓥)
+                → (⟪ B ⟫ → ⟨ L ⟩) → Ω 𝓤
+is-generated-by {𝓦 = 𝓦} L B η =
+ Ɐ x ∶ ⟨ L ⟩ , x ＝[ σ ]＝ (⋁[ L ] ⁅ η b ∣ (b , _) ∶ (Σ b ꞉ ⟪ B ⟫ , η b ≤ x) ⁆)
+  where
+   σ : is-set ⟨ L ⟩
+   σ = carrier-of-[ poset-of L ]-is-set
+
+   _≤_ = λ x y → (x ≤[ poset-of L ] y) holds
+
+contains-compact-opens : (L : Frame 𝓤 𝓦 𝓦) (B : BooleanAlgebra 𝓦 𝓥)
+                       → (⟪ B ⟫ → ⟨ L ⟩) → Ω (𝓤 ⊔ 𝓦 ⁺)
+contains-compact-opens L B η =
+ Ɐ x ∶ ⟨ L ⟩ , is-compact-open L x ⇒ (Ǝ b ∶ ⟪ B ⟫ , η b ＝ x)
+
+\end{code}
+
+\begin{code}
+
+extension-lemma : (B : BooleanAlgebra 𝓦 𝓥) (L L′ : Frame 𝓤 𝓦 𝓦)
+                → (η : ⟪ B ⟫ → ⟨ L ⟩)
+                → is-embedding B L η holds
+                → is-spectral L holds
+                → is-spectral L′ holds
+                → is-spectral′ B L η holds
+                → is-generated-by L B η holds
+                → contains-compact-opens L B η holds
+                → (h : ⟪ B ⟫ → ⟨ L′ ⟩)
+                → is-lattice-homomorphism B L′ h holds
+                → is-contr
```

Comment:
`is-contr (\Sigma ...)` can be abbreviated as `\exists! ...`.

Reply:
Done.